scripts/linking.py

```diff
@@ -405,33 +421,94 @@ def get_mention_text_and_ids_by_doc(data: List[data_util.MedMentionExample],
         A dictionary of UMLS concepts
     """
     missing_entity_ids = []  # entities in MedMentions but not in UMLS
-
     examples_with_labels = []
 
+    substituted = 0
+    total = 0
+
     for example in data:
+
+        doc = nlp(example.text)
+        abbreviations = {}
+        if substitute_abbreviations:
+            for full, shorts in doc._.abbreviations:
+                for short in shorts:
+                    abbreviations[short] = full
+
         mention_texts = []
+        predicted_mention_texts = []
         gold_umls_ids = []
         for entity in example.entities:
             if entity.umls_id not in umls:
                 missing_entity_ids.append(entity)  # the UMLS release doesn't contan all UMLS concepts
                 continue
 
-            mention_texts.append(entity.mention_text)
+            _, mention_string = maybe_substitute_span(doc, entity, abbreviations)
+            if mention_string is None:
+                mention_string = entity.mention_text
+
+            if mention_string != entity.mention_text:
+                substituted += 1
+            mention_texts.append(mention_string)
+
             gold_umls_ids.append(entity.umls_id)
-            continue
-        examples_with_labels.append((example, mention_texts, gold_umls_ids))
+            total += 1
+
+        for entity in doc.ents:
+            new_span, _ = maybe_substitute_span(doc, entity, abbreviations)
+            if new_span is None:
+                predicted_mention_texts.append(entity)
+            else:
+                # We have to manually create a new span with the new start and end points, but with the old label,
+                # as spans are read only views of a document.
+                span_with_label = Span(doc, start=new_span.start, end=new_span.end, label=entity.label_)
+                predicted_mention_texts.append(span_with_label)
```

Comment:
do you need to also remove the old entity span here?

Reply:
Looking at this again, I think it doesn't introduce any bugs, but might be a bit confusing to have more entities on the doc than on the MedMentionsExample, but I don't think it matters

Reply:
I added a comment that they might not match - it's ok because we don't use the `doc.ents` in the main function anymore (but you are correct that it's important that we don't, because it would mean the abbreviations are not included regardless of the flag.)